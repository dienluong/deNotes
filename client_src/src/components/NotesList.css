/* NotesList styles */
.active-tree-node {
  box-shadow: 0 0 0 4px pink;
}
