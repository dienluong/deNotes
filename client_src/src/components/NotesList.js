/* TODO: Revisit this component and react-ui-tree code to see if rendering can be optimized
   https://reactjs.org/docs/optimizing-performance.html
*/
// import 'react-ui-tree/dist/react-ui-tree.css';
// import Tree from 'react-ui-tree';
// TODO: Delete Note.js
// import Note from './Note';

import Tree, { removeNode } from 'react-sortable-tree';
import 'react-sortable-tree/style.css';
// TODO: Delete NotesList.css
// import './NotesList.css';
import React from 'react';

import sampleNotes from '../test/sample-tree';

class NotesList extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      notesTree: sampleNotes,
    };

    this.handleChange = this.handleChange.bind(this);
    // this.renderNode = this.renderNode.bind(this);
    // this.onClickNode = this.onClickNode.bind(this);
    this.deleteNote = this.deleteNote.bind(this);
  }

  handleChange(notesTree) {
    this.setState({
      notesTree,
    });
  }

  /* TODO: Remove, formerly for react-ui-tree
  renderNode(node) {
    // console.log(`renderNode ${node.module}`);
    return (
      <Note
        active={ this.state.activeNote === node }
        clickHandler={ this.onClickNode }
        node={ node }
      />
    );
  }

  onClickNode(node) {
    // console.log(`Active Note: ${node.module}`);
    this.setState({
      activeNote: node,
    });
  }
  */

  deleteNote({ path }) {
    return ({
      buttons: [
        <button
          style={{ verticalAlign: 'middle' }}
          onClick={() => {
            const { treeData } = removeNode({
              treeData: this.state.notesTree,
              getNodeKey: ({ treeIndex }) => treeIndex,
              path,
            });
            this.setState({
              notesTree: treeData,
            });
          }}
        >
          x
        </button>,
      ],
    });
  }

  render() {
    console.log('rendering NoteLists........................');
    return (
      <Tree
        className='tree'
        treeData={ this.state.notesTree }
        onChange={ this.handleChange }
        generateNodeProps={ this.deleteNote }
      />
    );
  };
}

export default NotesList;
