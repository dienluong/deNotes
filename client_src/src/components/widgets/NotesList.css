/* NotesList styles */
.dnt__active-tree-node {
  box-shadow: 0 0 0 4px pink;
}

.dnt__tree-node-btn {
  vertical-align: middle;
}
