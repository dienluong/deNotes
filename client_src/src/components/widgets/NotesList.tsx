import React from 'react';
import Toolbar from './Toolbar';
import PathNavigator from './PathNavigator';
import Tree from 'react-sortable-tree';
import mobileTheme from '../../tree-theme';
import 'react-sortable-tree/style.css';
import styles from './NotesList.module.css';
import NodeTitle from './NodeTitle';

// Types
import { TreeItem } from 'react-sortable-tree';
type PropsT = {
  tree: TreeNodeT[],
  activeNode: ActiveNodeT,
  activePath: string[],
  treeChangeHandler: (...args: any) => any;
  nodeTitleChangeHandler: (...args: any) => any,
  nodeClickHandler: (params: { id: TreeNodeT["id"], path: TreeNodePathT }) => unknown,
  nodeDoubleClickHandler: (params: { id: TreeNodeT["id"], path: TreeNodePathT }) => unknown,
  deleteNodeBtnHandler: (params: { node: TreeNodeT, path: TreeNodePathT }) => unknown,
  addNoteBtnHandler: (params: { path: TreeNodePathT }) => unknown,
  pathNavigatorClickHandler: (...args: any) => any,
  toolbarHandlersMap: any,
  getNodeKey: (...args: any) => any,
};
type generateNodePropsT = ({ node, path }: { node: TreeItem, path: (string|number)[] }) => object;

function NotesList({
  tree,
  activeNode,
  activePath,
  treeChangeHandler,
  nodeTitleChangeHandler,
  nodeClickHandler,
  nodeDoubleClickHandler,
  deleteNodeBtnHandler,
  addNoteBtnHandler,
  pathNavigatorClickHandler,
  toolbarHandlersMap,
  getNodeKey,
}: PropsT) {
  function buildNodeProps({ node, path }: { node: TreeNodeT, path: string[] }) {
    return ({
      title: (<NodeTitle node={ node } path={ path } onSubmit={ nodeTitleChangeHandler } />),
      className: (node.id === activeNode.id) ? `${styles['dnt__tree-node']} ${styles['dnt__tree-node--active']}` : styles['dnt__tree-node'],
      buttons: _buildNodeButtons({ node, path }),
      tabIndex: '0',
      'data-testid': node.id,
      onClick: () => nodeClickHandler({ id: node.id, path }),
      onDoubleClick: () => nodeDoubleClickHandler({ id: node.id, path }),
    });
  }

  function _buildNodeButtons({ node, path }: { node: TreeNodeT, path: string[] }) {
    let buttons = [
      <button
        className={ styles['dnt__tree-node-btn'] }
        onClick={ (event) => {
          event.stopPropagation();
          deleteNodeBtnHandler({ node, path });
        }}
      >
        x
      </button>,
    ];

    // Check if current node is parent node; '+' button only for parent nodes
    if (typeof node.children !== 'undefined') {
      buttons.unshift(
        <button
          className={ styles['dnt__tree-node-btn'] }
          onClick={ (event) => {
            event.stopPropagation();
            addNoteBtnHandler({ path });
          }}
        >
          +
        </button>);
    }

    return buttons;
  }

  function _collapseAllParents({ tree }: { tree: TreeNodeT[] }): TreeNodeT[] {
    return tree.map((node): TreeNodeT => {
      if (node && node.children) {
        return { ...node, expanded: false };
      } else {
        return node;
      }
    });
  }

  return (
    <div className={ styles["dnt__notes-list"] }>
      <Toolbar toolsMap={ toolbarHandlersMap } />
      <PathNavigator
        path={ activePath }
        activeSegmentIdx={ activeNode.path.indexOf(activeNode.id) }
        onClick={ pathNavigatorClickHandler }
      />
      <Tree
        className={ 'tree ' + styles.dnt__tree }
        treeData={ _collapseAllParents({ tree }) }
        theme={ mobileTheme as any }
        onChange={ treeChangeHandler }
        getNodeKey={ getNodeKey }
        generateNodeProps={ buildNodeProps as generateNodePropsT }
      />
    </div>
  );
}

export default NotesList;
