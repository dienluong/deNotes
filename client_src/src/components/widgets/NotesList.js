import React, { Fragment } from 'react';
import Toolbar from './Toolbar';
import PathNavigator from './PathNavigator';
import Tree from 'react-sortable-tree';
import 'react-sortable-tree/style.css';
import './NotesList.css';
import NodeTitle from './NodeTitle';

function NotesList({
  notesTree,
  activeNode,
  activePath,
  treeChangeHandler,
  nodeTitleChangeHandler,
  nodeClickHandler,
  deleteNodeBtnHandler,
  addNoteBtnHandler,
  pathNavigatorClickHandler,
  toolbarHandlersMap,
  getNodeKey,
}) {
  function buildNodeProps({ node, path }) {
    return ({
      title: (<NodeTitle node={ node } path={ path } onSubmit={ nodeTitleChangeHandler } />),
      className: (node.id === activeNode.id) ? 'dnt__active-tree-node' : '',
      buttons: _buildNodeButtons({ node, path }),
      tabIndex: '0',
      onClick: () => nodeClickHandler({ id: node.id, path }),
      onFocus: () => console.log('Focused: ' + node.title),
      onBlur: () => console.log('unFocused: ' + node.title),
    });
  }

  function _buildNodeButtons({ node, path }) {
    let buttons = [
      <button
        className='dnt__tree-node-btn'
        onClick={ (event) => {
          event.stopPropagation();
          deleteNodeBtnHandler({ node, path });
        }}
      >
        x
      </button>,
    ];

    // Check if current node is parent node; '+' button only for parent nodes
    if (typeof node.children !== 'undefined') {
      buttons.unshift(
        <button
          className='dnt__tree-node-btn'
          onClick={ (event) => {
            event.stopPropagation();
            addNoteBtnHandler({ path });
          }}
        >
          +
        </button>);
    }

    return buttons;
  }

  // TODO: remove
  // console.log(`
  //   Tree: ${JSON.stringify(notesTree)} \n
  //   Active ID: ${activeNode.id} \n
  //   Path: ${activeNode.path} \n
  //   ${ activePath }
  // `);

  return (
    <Fragment>
      <Toolbar toolsMap={ toolbarHandlersMap } />
      <PathNavigator
        path={ activePath }
        activeSegmentIdx={ activeNode.path.indexOf(activeNode.id) }
        onClick={ pathNavigatorClickHandler }
      />
      <Tree
        className='tree'
        treeData={ notesTree }
        onChange={ treeChangeHandler }
        getNodeKey={ getNodeKey }
        generateNodeProps={ buildNodeProps }
      />
    </Fragment>
  );
}

export default NotesList;
