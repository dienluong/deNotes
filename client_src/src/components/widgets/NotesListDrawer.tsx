import React from 'react';
import Drawer from '@material-ui/core/Drawer';
import Hidden from '@material-ui/core/Hidden';
import NotesList from './NotesList';
import styles from './NotesListDrawer.module.css';

function NotesListDrawer({ drawerOpen, handleDrawerToggle, ...otherProps }: any) {
  // const [mobileOpen, setMobileOpen] = React.useState(false);
  // function handleDrawerToggle() {
  //   setMobileOpen(!mobileOpen);
  // }

  // TODO remove
  console.log('drawerOpen]]]]] ', drawerOpen);

  return (
    <div className={ styles['dnt__notes-list-drawer-root'] }>
    <nav className={ styles['dnt__notes-list-drawer'] }>
      <Hidden smUp implementation="css">
        <Drawer
          variant="temporary"
          anchor="left"
          open={ drawerOpen }
          onClose={ handleDrawerToggle }
          ModalProps={{
            keepMounted: true, // Better open performance on mobile.
          }}
          classes={{
            paper: styles['dnt__notes-list-drawer-paper'],
          }}
        >
          <NotesList {...otherProps} />
        </Drawer>
      </Hidden>
      <Hidden xsDown implementation="css">
        <Drawer
          variant="permanent"
          anchor="left"
          open
          classes={{
            paper: styles['dnt__notes-list-drawer-paper'],
          }}
        >
          <NotesList {...otherProps} />
        </Drawer>
      </Hidden>
    </nav>
    </div>
  );
}

export default NotesListDrawer;
