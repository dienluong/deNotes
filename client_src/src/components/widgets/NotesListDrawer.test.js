import React from 'react';
import NotesListDrawer from './NotesListDrawer';
import NotesList from './NotesList';
import Drawer from '@material-ui/core/Drawer';
import { createSerializer } from 'enzyme-to-json';

// Configure Enzyme
import { configure, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
configure({ adapter: new Adapter() });
let wrapper = {};

// Configure snapshot serializer
expect.addSnapshotSerializer(createSerializer());

afterEach(() => {
  if (typeof wrapper.unmount === 'function') {
    wrapper.unmount();
  }
});

it('should render drawer using material-ui Drawer w/ proper props and pass remaining props to NotesList', () => {
  const handleDrawerToggle = jest.fn();
  const drawerCloseHandler = jest.fn();
  const notesListProp1 = 'dummy-prop';
  const notesListProp2 = 1234;
  const props = {
    drawerOpen: false,
    drawerSide: 'right',
    size: 'small',
    handleDrawerToggle,
    drawerCloseHandler,
    notesListProp1,
    notesListProp2,
  };
  const expectedDrawerProps = {
    anchor: props.drawerSide,
    open: props.drawerOpen,
  };
  const expectedNotesListProps = {
    notesListProp1,
    notesListProp2,
  };

  wrapper = shallow(<NotesListDrawer { ...props } />);
  expect(wrapper).toMatchSnapshot();
  const drawerWrapper = wrapper.find(Drawer);
  expect(drawerWrapper.exists()).toBe(true);
  expect(drawerWrapper.props()).toMatchObject(expectedDrawerProps);
  const notesListWrapper = drawerWrapper.find(NotesList);
  expect(notesListWrapper.exists()).toBe(true);
  expect(notesListWrapper.props()).toMatchObject(expectedNotesListProps);
});

it('should call expected handlers on close event', () => {
  const handleDrawerToggle = jest.fn();
  const drawerCloseHandler = jest.fn();
  const notesListProp1 = 'dummy-prop';
  const notesListProp2 = 1234;
  const props = {
    drawerOpen: false,
    drawerSide: 'right',
    size: 'small',
    handleDrawerToggle,
    drawerCloseHandler,
    notesListProp1,
    notesListProp2,
  };

  const drawerWrapper = shallow(<NotesListDrawer { ...props } />).find(Drawer);
  expect(drawerWrapper.exists()).toBe(true);
  const onCloseHandler = drawerWrapper.prop('onClose');
  expect(onCloseHandler).toBeInstanceOf(Function);
  onCloseHandler();
  expect(handleDrawerToggle).toBeCalledTimes(1);
  expect(drawerCloseHandler).toBeCalledTimes(1);
});
