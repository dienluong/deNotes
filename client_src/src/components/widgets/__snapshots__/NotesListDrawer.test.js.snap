// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render drawer using material-ui Drawer w/ proper props and pass remaining props to NotesList 1`] = `
<Fragment>
  <WithStyles(Drawer)
    ModalProps={
      Object {
        "keepMounted": true,
      }
    }
    anchor="right"
    classes={
      Object {
        "paper": "dnt__notes-list-drawer-paper--small",
      }
    }
    onClose={[Function]}
    open={false}
    variant="temporary"
  >
    <NotesList
      notesListProp1="dummy-prop"
      notesListProp2={1234}
      size="small"
    />
  </WithStyles(Drawer)>
  <Connect(ModalManager) />
</Fragment>
`;
